src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,634 @@
+    std::map<vertex_t, std::set<vertex_t>> result = {};
+    BGL_FORALL_VERTICES_T(v, pattern, graph_t) {
+        for (Embedding const& embedding : embeddings) {
+            std::set<vertex_t> vertices = {embedding.at(v)};
+            auto [it, emplaced] = result.try_emplace(v, std::move(vertices));
+            if (!emplaced) {
+                it->second.insert(embedding.at(v));
+            }
+        }
+    }
+
+    std::size_t min_size = boost::num_vertices(graph);
+    for (auto const& maps : result) {
+        min_size = std::min(maps.second.size(), min_size);
+    }
+    return min_size;
+}
+
+bool IsForbidden(std::vector<Rules>::iterator const& forbidden_rules,
+                 std::vector<Literal> const& subset, Literal const& l) {
+    for (auto const& [premises, conclusion] : *forbidden_rules) {
+        if (CompareLiteralSets(premises, subset) && ContainsLiteral(conclusion, l)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void HGenerate(std::vector<Literal>& notsatisfied, Literal const& l, std::vector<SingleRule>& rules,
+               graph_t const& graph, graph_t& pattern, Embeddings const& embeddings,
+               std::vector<Rules>::iterator const& forbidden_rules, std::size_t const sigma) {
+    std::vector<Literal> conclusion = {l};
+    for (std::size_t k = 1; k <= 3; ++k) {
+        std::vector<std::vector<Literal>> subsets = GetSubsets<Literal>(notsatisfied, k);
+        for (auto& subset : subsets) {
+            if (IsForbidden(forbidden_rules, subset, l)) {
+                continue;
+            }
+
+            Gfd gfd = {pattern, subset, conclusion};
+            if (Validate(graph, gfd, embeddings, sigma)) {
+                rules.emplace_back(subset, l);
+                for (auto const& lit : subset) {
+                    auto last = std::remove(notsatisfied.begin(), notsatisfied.end(), lit);
+                    notsatisfied.erase(last, notsatisfied.end());
+                }
+            }
+        }
+    }
+}
+
+std::vector<SingleRule> GenerateRules(std::vector<Literal> const& literals, graph_t const& graph,
+                                      graph_t& pattern, Embeddings const& embeddings,
+                                      std::vector<Rules>::iterator const& forbidden_rules,
+                                      std::size_t const sigma) {
+    std::vector<SingleRule> rules = {};
+    for (std::size_t j = 0; j < literals.size(); ++j) {
+        std::vector<Literal> current = literals;
+        Literal l = current.at(j);
+        current.erase(std::next(current.begin(), j));
+
+        if (IsForbidden(forbidden_rules, {}, l)) {
+            continue;
+        }
+
+        std::vector<Literal> conclusion = {l};
+        std::vector<Literal> premises = {};
+        Gfd gfd = {pattern, premises, conclusion};
+        if (Validate(graph, gfd, embeddings, sigma)) {
+            rules.push_back({{}, l});
```

Comment:
To avoid unnecessary temporary object creating
```suggestion
            rules.emplace_back({}, l);
```
Consider everywhere else

Reply:
the suggested line gives a compiler error:
no matching function for call to ‘std::vector<std::pair<std::vector<std::pair<std::pair<int, std::__cxx11::basic_string<char> >, std::pair<int, std::__cxx11::basic_string<char> > > >, std::pair<std::pair<int, std::__cxx11::basic_string<char> >, std::pair<int, std::__cxx11::basic_string<char> > > > >::emplace_back(<brace-enclosed initializer list>, gfd::Literal&)’